src/psyclone/psyGen.py

```diff
@@ -1488,6 +1490,39 @@ def dag_name(self):
         '''
         return "ACC_parallel_" + str(self.abs_position)
 
+    def _pre_gen_validate(self):
+        '''
+        Check that the PSyIR tree containing this node is valid. Since we
+        use 'default(present)', this node must either be the child of an
+        ACCDataDirective or the parent Schedule must contain an
+        ACCEnterDataDirective before this one.
+
+        :raises GenerationError: if this ACCParallel node is not preceded by \
+            an ACCEnterDataDirective and is not the child of an \
+            ACCDataDirective.
+
+        '''
+        # We can't use Node.ancestor() because the enter data directive does
+        # not have children. Instead, we go back up to the Schedule and
+        # walk down from there.
+        routine = self.ancestor(Routine)
+        nodes = routine.walk(ACCEnterDataDirective)
+        # Check that any enter-data directive comes before this parallel
+        # directive
+        if nodes and nodes[0].abs_position > self.abs_position:
+            raise GenerationError(
```

Comment:
Codecov points to this line as not covered, can you check this exception.

Reply:
Oh yes, sorry I missed that. I wonder if it has always been uncovered? Took a bit of work but I've added a couple of new tests.